src/platforms/esp32/components/avm_builtins/storage_nif.c

```diff
@@ -0,0 +1,336 @@
+        VALIDATE_VALUE(cs_term, term_is_integer);
+        spi_slot_config.gpio_cs = term_to_int(cs_term);
+
+        term cd_term = interop_kv_get_value_default(
+            opts_term, ATOM_STR("\x2", "cd"), UNDEFINED_ATOM, ctx->global);
+        if (cd_term != UNDEFINED_ATOM) {
+            VALIDATE_VALUE(cd_term, term_is_integer);
+            spi_slot_config.gpio_cd = term_to_int(cd_term);
+        }
+
+        mount = enif_alloc_resource(platform->mounted_fs_resource_type, sizeof(struct MountedFS));
+        if (IS_NULL_PTR(mount)) {
+            RAISE_ERROR(OUT_OF_MEMORY_ATOM);
+        }
+        mount->base_path = strdup(target);
+        mount->mount_type = FATSPIFlash;
+
+        ret = esp_vfs_fat_sdspi_mount(
+            target, &host_config, &spi_slot_config, &mount_config, &mount->handle.card);
+    } else {
+        RAISE_ERROR(BADARG_ATOM);
+    }
+
+    term return_term = term_invalid_term();
+    if (UNLIKELY(ret != ESP_OK)) {
+        mount->mount_type = Unmounted;
+        return_term = make_esp_error_tuple(ret, ctx);
+    } else {
+        if (UNLIKELY(memory_ensure_free(ctx, TUPLE_SIZE(2) + TERM_BOXED_RESOURCE_SIZE) != MEMORY_GC_OK)) {
+            RAISE_ERROR(OUT_OF_MEMORY_ATOM);
+        }
+        term mount_term = enif_make_resource(erl_nif_env_from_context(ctx), mount);
+        term return_term = term_alloc_tuple(2, &ctx->heap);
+        term_put_tuple_element(return_term, 0, OK_ATOM);
+        term_put_tuple_element(return_term, 1, mount_term);
+    }
+    enif_release_resource(mount);
+
+    free(source);
+    free(target);
+
+    return return_term;
+}
+
+static esp_err_t do_umount(struct MountedFS *mount)
+{
+    esp_err_t ret = ESP_FAIL;
+
+    switch (mount->mount_type) {
+        case Unmounted:
+            ret = ESP_OK;
+            break;
+
+        case FATSPIFlash:
+#if ESP_IDF_VERSION_MAJOR >= 5
+            ret = esp_vfs_fat_spiflash_unmount_rw_wl(mount->base_path, mount->handle.wl);
+#else
+            ret = esp_vfs_fat_spiflash_unmount(mount->base_path, mount->handle.wl);
+#endif
+            break;
+
+        case FATSDSPI:
+        case FATSDMMC:
+            ret = esp_vfs_fat_sdcard_unmount(mount->base_path, mount->handle.card);
+            break;
+    }
+
+    return ret;
+}
+
+static term nif_esp_umount(Context *ctx, int argc, term argv[])
+{
+    GlobalContext *glb = ctx->global;
+    struct ESP32PlatformData *platform = glb->platform_data;
+
+    void *mount_obj_ptr;
+    if (UNLIKELY(!enif_get_resource(erl_nif_env_from_context(ctx), argv[0], platform->mounted_fs_resource_type, &mount_obj_ptr))) {
+        RAISE_ERROR(BADARG_ATOM);
+    }
+    struct MountedFS *mounted_fs = (struct MountedFS *) mount_obj_ptr;
+
+    if (UNLIKELY(mounted_fs->mount_type == Unmounted)) {
+        RAISE_ERROR(BADARG_ATOM);
+    }
+
+    esp_err_t ret = do_umount(mounted_fs);
+
+    if (UNLIKELY(ret != ESP_OK)) {
+        return make_esp_error_tuple(ret, ctx);
+    }
+
+    mounted_fs->mount_type = Unmounted;
```

Comment:
@pguyot I'm looking for a suggestion here, should I protect the MountedFS resource with some kind of lock?

Reply:
After analysis, I believe we need some kind of protection because esp idf function `esp_vfs_fat_spiflash_unmount_rw_wl` doesn't look thread safe to me.

https://github.com/espressif/esp-idf/blob/46acfdce969f03c02b001fe4d24fa9e98f6adc5e/components/fatfs/vfs/vfs_fat_spiflash.c#L214

If we call it twice with the same value close enough, first core would free the pointer and second would try to free it before it's nullified.

The lock could be implemented around mount_type with a special umounting type, or just setting umounted before it's actually umounted, since we don't have the umounted => mounted transition.

Reply:
At the end of the day I realized that using a SpinLock was  guaranteed to be safe and headache free.